specs/interop/storage-upgrade.md

```diff
@@ -0,0 +1,79 @@
+# Migrated Liquidity storage update
+
+## Overview
+
+To fully migrate liquidity from `OptimismMintableERC20` tokens to `OptimismSuperchainERC20`, the `OptimismMintableERC20Factory` needs to track its deployment history.
+
+To achieve this, a [Hash Onion](https://github.com/ethereum-optimism/design-docs/blob/main/protocol/superchain-erc20/storage-upgrade.md#2-hash-onion) solution will be used:
+
+- The `OptimismMintableERC20Factory` predeploy will include functionality to set and store the Hash Onion, as well as verify and record deployment data.
+- A Foundry script will be added to generate the Hash Onion from a list of local and remote tokens.
+- A `superchain-ops` task will be added to set the Hash Onion from the L2 `ProxyAdmin` owner.
+
+## `OptimismMintableERC20Factory` predeploy
+
+Create a new `OptimismMintableERC20FactoryInterop` contract that inherits from `OptimismMintableERC20Factory` and add the following functionality:
+
+```solidity
+bytes32 public onionHash;
+
+function setOnionHash(bytes32 _onionHash) external {
+    if (msg.sender != Predeploys.PROXY_ADMIN) revert("Unauthorized");
+    if (hashOnion != 0) revert("Already initialized");
+
+    onionHash = _onionHash;
+}
+
+function verifyAndStore(
+	address localToken,
+	address remoteToken,
+	bytes32 innerLayer
+) public {
+	// Concatenate localToken and remoteToken using abi.encodePacked
+	bytes memory data = abi.encodePacked(localToken, remoteToken);
+
+	// Recompute the hash of the previous layer + data
+	bytes32 computedHash = keccak256(abi.encodePacked(innerLayer, data));
+
+	// Ensure the computed hash matches the current stored layer
+	require(computedHash == hashOnion, "Invalid layer provided.");
+
+	// Store the localToken and remoteToken in the mapping
+	deployments[localToken] = remoteToken;
+
+	// Emit event for logging
+	emit AddressesStored(localToken, remoteToken);
+
+	// Update the current layer to the previous one for the next submission
+	onionHash = innerLayer;
+}
+```
+
+Also a function that receives local and remote tokens arrays can be implemented so that many layers can be verified and stored together.
+
+## Generate Hash Onion
+
+Create a Foundry script that generate the final onion hash based on a list of local and remote tokens.
+
+```solidity
+function generateOnionHash(
+	address[] calldata localTokens,
+    address[] calldata remoteTokens
+) public {
+    require(localTokens.length == remoteTokens.length, "Invalid arrays");
+
+    bytes32 onionHash = keccak256(0);
```

Comment:
not sure how does foundry handle this. We just discussed wheter to do keccak256("0") or keccak256(abi.encode(0)). I think the latter is better for consistency with the rest